pkg/synth/acl.go

```diff
@@ -15,153 +15,177 @@ import (
 	for _, src := range conn.Src.IPAddrs {
+		srcCidrs, src := adjustResource(s, src, conn.Src)
 		for _, dst := range conn.Dst.IPAddrs {
+			dstCidrs, dst := adjustResource(s, dst, conn.Dst)
 			if src == dst && conn.Src.Type != ir.ResourceTypeCidr && conn.Dst.Type != ir.ResourceTypeCidr {
 				continue
 			}
 			for _, trackedProtocol := range conn.TrackedProtocols {
 				reason := explanation{internal: internal, connectionOrigin: conn.Origin, protocolOrigin: trackedProtocol.Origin}
-				protocolRules := allowDirectedConnection(s, src, dst, conn, internalSrc, internalDst, trackedProtocol.Protocol, reason)
-				connectionRules = append(connectionRules, protocolRules...)
+				allowDirectedConnection(src, dst, srcCidrs, dstCidrs, internalSrc, internalDst, trackedProtocol.Protocol, reason, result, single)
 			}
 		}
 	}
-	for _, rule := range connectionRules {
-		acl := result.LookupOrCreate(aclSelector(rule.Target()))
-		if internal {
-			acl.AppendInternal(rule)
-		} else {
-			acl.AppendExternal(rule)
-		}
-	}
 	return result
 }
 
-func allowDirectedConnection(s *ir.Spec, srcCidr, dstCidr *ipblock.IPBlock, conn *ir.Connection, internalSrc, internalDst bool,
-	protocol ir.Protocol, reason explanation) []*ir.ACLRule {
+// create relevant rule(s) to allow traffic
+func allowDirectedConnection(srcCidr, dstCidr *ipblock.IPBlock, srcSubnets, dstSubnets []*ir.ConnResource, internalSrc, internalDst bool,
+	protocol ir.Protocol, reason explanation, result *ir.ACLCollection, singleACL bool) {
 	var request, response *ir.Packet
-
-	srcCidr = expandNifToSubnet(&s.Defs.ConfigDefs, &conn.Src, srcCidr)
-	dstCidr = expandNifToSubnet(&s.Defs.ConfigDefs, &conn.Dst, dstCidr)
-
-	srcSubnetList := subnetsContainedInCidr(s, srcCidr, conn.Src)
-	dstSubnetList := subnetsContainedInCidr(s, dstCidr, conn.Dst)
-
-	var connection []*ir.ACLRule
+	internal := internalSrc && internalDst
 
 	if internalSrc {
-		for _, srcSubnetCidr := range srcSubnetList {
-			if srcSubnetCidr == dstCidr {
+		for _, srcSubnet := range srcSubnets {
+			if srcSubnet.Addrs.Equal(dstCidr) { // srcSubnet and dstCidr are the same subnet
 				continue
 			}
-			request = &ir.Packet{Src: srcSubnetCidr, Dst: dstCidr, Protocol: protocol, Explanation: reason.String()}
-			connection = append(connection, ir.AllowSend(*request))
+			request = &ir.Packet{Src: srcSubnet.Addrs, Dst: dstCidr, Protocol: protocol, Explanation: reason.String()}
+			addRuleToACL(result, ir.AllowSend(*request), srcSubnet, internal, singleACL)
 			if inverseProtocol := protocol.InverseDirection(); inverseProtocol != nil {
-				response = &ir.Packet{Src: dstCidr, Dst: srcSubnetCidr, Protocol: inverseProtocol, Explanation: reason.response().String()}
-				connection = append(connection, ir.AllowReceive(*response))
+				response = &ir.Packet{Src: dstCidr, Dst: srcSubnet.Addrs, Protocol: inverseProtocol, Explanation: reason.response().String()}
+				addRuleToACL(result, ir.AllowReceive(*response), srcSubnet, internal, singleACL)
```

Comment:
There is no point in taking the pointer of a struct and then only using it for dereferencing.
Better change `AllowReceive` and `AllowSend` (and `packetACLRule`) to take a pointer to `ir.Packet`.